File: module-8-harness/solutions/mcp-server.ts
/**
 * Module 8: Minimal MCP server
 * Exposes a custom tool over the Model Context Protocol.
 *
 * This is a simplified MCP server that responds to JSON-RPC over stdin/stdout.
 * In production, use the @modelcontextprotocol/sdk package.
 */
import * as readline from "readline";

interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: number | string;
  method: string;
  params?: unknown;
}

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id: number | string;
  result?: unknown;
  error?: { code: number; message: string };
}

const TOOLS = [
  {
    name: "get-project-info",
    description: "Returns information about the current project",
    inputSchema: {
      type: "object",
      properties: {},
      required: [],
    },
  },
];

function handleRequest(req: JsonRpcRequest): JsonRpcResponse {
  switch (req.method) {
    case "initialize":
      return {
        jsonrpc: "2.0",
        id: req.id,
        result: {
          protocolVersion: "2024-11-05",
          capabilities: { tools: {} },
          serverInfo: { name: "ai-agent-basics-mcp", version: "1.0.0" },
        },
      };

    case "tools/list":
      return { jsonrpc: "2.0", id: req.id, result: { tools: TOOLS } };

    case "tools/call": {
      const params = req.params as { name: string; arguments?: Record<string, unknown> };
      if (params.name === "get-project-info") {
        return {
          jsonrpc: "2.0",
          id: req.id,
          result: {
            content: [
              {
                type: "text",
                text: JSON.stringify({
                  name: "AI Agent Basics",
                  modules: 12,
                  language: "TypeScript",
                  description: "Hands-on curriculum for building coding agents from scratch",
                }),
              },
            ],
          },
        };
      }
      return {
        jsonrpc: "2.0",
        id: req.id,
        error: { code: -32601, message: `Unknown tool: ${params.name}` },
      };
    }

    default:
      return {
        jsonrpc: "2.0",
        id: req.id,
        error: { code: -32601, message: `Unknown method: ${req.method}` },
      };
  }
}

// Read JSON-RPC messages from stdin
const rl = readline.createInterface({ input: process.stdin });
rl.on("line", (line) => {
  try {
    const req: JsonRpcRequest = JSON.parse(line);
    const res = handleRequest(req);
    process.stdout.write(JSON.stringify(res) + "\n");
  } catch {
    // Ignore malformed input
  }
});

console.error("MCP server started (stdin/stdout)");

File: module-8-harness/solutions/python/mcp_server.py
"""Module 8: Minimal MCP server (stdin/stdout JSON-RPC)."""
import sys
import json

TOOLS = [
    {
        "name": "get-project-info",
        "description": "Returns information about the current project",
        "inputSchema": {"type": "object", "properties": {}, "required": []},
    },
]


def handle_request(req: dict) -> dict:
    method = req.get("method", "")

    if method == "initialize":
        return {
            "jsonrpc": "2.0", "id": req["id"],
            "result": {
                "protocolVersion": "2024-11-05",
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "ai-agent-basics-mcp", "version": "1.0.0"},
            },
        }

    if method == "tools/list":
        return {"jsonrpc": "2.0", "id": req["id"], "result": {"tools": TOOLS}}

    if method == "tools/call":
        params = req.get("params", {})
        if params.get("name") == "get-project-info":
            return {
                "jsonrpc": "2.0", "id": req["id"],
                "result": {
                    "content": [{
                        "type": "text",
                        "text": json.dumps({
                            "name": "AI Agent Basics",
                            "modules": 12,
                            "language": "TypeScript + Python",
                            "description": "Hands-on curriculum for building coding agents from scratch",
                        }),
                    }],
                },
            }
        return {"jsonrpc": "2.0", "id": req["id"], "error": {"code": -32601, "message": f"Unknown tool: {params.get('name')}"}}

    return {"jsonrpc": "2.0", "id": req.get("id", 0), "error": {"code": -32601, "message": f"Unknown method: {method}"}}


def main():
    print("MCP server started (stdin/stdout)", file=sys.stderr)
    for line in sys.stdin:
        try:
            req = json.loads(line.strip())
            res = handle_request(req)
            print(json.dumps(res), flush=True)
        except Exception:
            pass


if __name__ == "__main__":
    main()
